Add I/O lab (#30)
open-education-hub/operating-systems: content/chapters/io/lab/content/redirections.md

# Redirections

> In the [previous section](./file-descriptors.md), we mentioned redirections such as `ls > file.txt`.

Reviewer: ```suggestion
In the [file descriptors](./file-descriptors.md), we mentioned redirections such as `ls > file.txt`.
```

## Practice: Naive Redirection

But before diving into reimplementing shell functionalities, let's look at a simpler example.
Navigate to `support/redirect/redirect.c`.
The code makes a naive attempt at redirecting the newly opened file to `stdout`.
It simply closes `stdout` first so that when `open()` returns **the lowest available file descriptor**, that value will be 1, which is `STDOUT_FILENO`.

Note there's a difference between `stdout` and `STDOUT_FILENO`.
While `stdout` is of type `FILE *` and is meant to be used with libc functions such as `fread()`, `STDOUT_FILENO` is the default file descriptor for the standard output, which almost always is 1.
So `STDOUT_FILENO` is an `int` type with the value 1.
Don't confuse them!

Compile and run the code without modifying it.
It pauses after each file descriptor operation.
In another terminal (or in another `tmux` window), run the following each time you press Enter in the first terminal/window: `lsof -p $(pidof redirect)`.

`lsof` displays the files opened by the given process.

> From the outpu below we see that these files are mainly files (opened as file descriptors) and libraries, which are memory mapped (`mem`).

Reviewer: ```suggestion
From the output below we see that these files are mainly files (opened as file descriptors) and libraries, which are memory mapped (`mem`).
```

On the third column, you can see the file descriptor corresponding to each file.

```
student@os:~/.../lab/support/redirect$ lsof -w -p $(pidof redirect)  # before any file operations
COMMAND    PID USER   FD   TYPE DEVICE SIZE/OFF    NODE NAME

> redirect 44303  teo  cwd    DIR    8,1     4096  299870 /media/teo/2TB/Chestii/Poli/Asistent/SO/operating-systems-oer/content/chapters/io/lab/support/redirect
> redirect 44303  teo  rtd    DIR  259,6     4096       2 /
> redirect 44303  teo  txt    REG    8,1    25848  299929 /media/teo/2TB/Chestii/Poli/Asistent/SO/operating-systems-oer/content/chapters/io/lab/support/redirect/redirect
> redirect 44303  teo  mem    REG  259,6  2029592 1857435 /usr/lib/x86_64-linux-gnu/libc-2.31.so
> redirect 44303  teo  mem    REG  259,6   191504 1835092 /usr/lib/x86_64-linux-gnu/ld-2.31.so
> redirect 44303  teo    0u   CHR  136,0      0t0       3 /dev/pts/0
> redirect 44303  teo    1u   CHR  136,0      0t0       3 /dev/pts/0
> redirect 44303  teo    2u   CHR  136,0      0t0       3 /dev/pts/0

Reviewer: Use generic names and paths.
```suggestion
redirect 44303  student  cwd    DIR    8,1     4096  299870 /home/student/operating-systems-oer/content/chapters/io/lab/support/redirect
redirect 44303  student  rtd    DIR  259,6     4096       2 /
redirect 44303  student  txt    REG    8,1    25848  299929 /home/student/operating-systems-oer/content/chapters/io/lab/support/redirect/redirect
redirect 44303  student  mem    REG  259,6  2029592 1857435 /usr/lib/x86_64-linux-gnu/libc-2.31.so
redirect 44303  student  mem    REG  259,6   191504 1835092 /usr/lib/x86_64-linux-gnu/ld-2.31.so
redirect 44303  student    0u   CHR  136,0      0t0       3 /dev/pts/0
redirect 44303  student    1u   CHR  136,0      0t0       3 /dev/pts/0
redirect 44303  student    2u   CHR  136,0      0t0       3 /dev/pts/0
```

```

Notice that all 3 default file descriptors are first liked to `/dev/pts/0`.
It may be different on your machine, but most likely it will be `/dev/pts/<some_number>`.
This is a [character device](./devices.md) that signifies your current (pseudo)terminal.

```console
student@os:~/.../lab/support/redirect$ lsof -w -p $(pidof redirect)  # after closing `STDOUT_FILENO`

> redirect 46906  teo    0u   CHR  136,0      0t0       3 /dev/pts/0
> redirect 46906  teo    2u   CHR  136,0      0t0       3 /dev/pts/0

Reviewer: ```suggestion
redirect 46906  student    0u   CHR  136,0      0t0       3 /dev/pts/0
redirect 46906  student    2u   CHR  136,0      0t0       3 /dev/pts/0
```

```

See that file descriptor 1 (`stdout`) has "disappeared".
Now the second entry in the process's FD table is free.

```console
student@os:~/.../lab/support/redirect$ lsof -w -p $(pidof redirect)  # after opening `redirect_file.txt`

> redirect 46906  teo    0u   CHR  136,0      0t0       3 /dev/pts/0
> redirect 46906  teo    1w   REG    8,1        0  299958 /.../lab/support/redirect/redirect_file.txt
> redirect 46906  teo    2u   CHR  136,0      0t0       3 /dev/pts/0

Reviewer: ```suggestion
redirect 46906  student    0u   CHR  136,0      0t0       3 /dev/pts/0
redirect 46906  student    1w   REG    8,1        0  299958 /.../lab/support/redirect/redirect_file.txt
redirect 46906  student    2u   CHR  136,0      0t0       3 /dev/pts/0
```

```

`open()` has assigned the newly opened file to the lowest file descriptor available, which is 1, the former `stdout`.
`printf()` writes its output to `stdout`, which in this case is `redirect_file.txt`.
Now inspect the contents of this file to make sure that string was written there.

[Quiz](../quiz/prints-work-no-stdio.md)

## Practice: Thread-unsafe Redirection

This is all fine, but it doesn't allow us to **copy** file descriptors.
We can only replace an existing file descriptor with another one.
To be able to do replacements, we can use the [`dup()`](https://man7.org/linux/man-pages/man2/dup.2.html) syscall.

> It simply creates a new fil descripto that refers to the same open file sturct as the one given to it as an argument.

Reviewer: ```suggestion
It simply creates a new file descriptor that refers to the same open file struct as the one given to it as an argument.
```

Change the `do_redirect()` function in `support/redirect/redirect.c` to employ this new logic.
It should follow the steps below.
Track them using `lsof`, just like before.

Step 1:

- 0 -> `stdin`
- 1 -> `stdout`
- 2 -> `stderr`

Step 2 - after `open("redirect_file.txt", ...)`:

- 0 -> `stdin`
- 1 -> `stdout`
- 2 -> `stderr`
- 3 -> `redirect_file.txt`

Step 3 - after `close(STDOUT_FILENO)`:

- 0 -> `stdin`
- 2 -> `stderr`
- 3 -> `redirect_file.txt`

Step 4 - after `dup(3)`.
Note that now both 1 and 3 are linked to `redirect_file.txt`, so we managed to successfully copy file descriptor 3.

- 0 -> `stdin`
- 1 -> `redirect_file.txt`
- 2 -> `stderr`
- 3 -> `redirect_file.txt`

Step 5 - after `close(3)`.
We don't need file descriptor 3 at this point anymore.

- 0 -> `stdin`
- 1 -> `redirect_file.txt`
- 2 -> `stderr`

## Practice: Safe Redirection

`dup()` is all fine and dandy, but what if 2 threads use the steps above concurrently?
Because steps 3 and 4 don't happen atomically, they risk having their results inverted.
Take a look at `support/redirect/redirect_parallel.c`.
Compile and run the code, then inspect the resulting files.

> You'll ntice they contain opposing strings:

Reviewer: ```suggestion
You'll notice they contain opposing strings:
```
